packages/eslint-plugin/src/rules/no-misused-object-likes.ts

```diff
@@ -0,0 +1,149 @@
+    checkObjectValuesForMap?: boolean;
+    checkObjectEntriesForMap?: boolean;
+    checkObjectKeysForSet?: boolean;
+    checkObjectValuesForSet?: boolean;
+    checkObjectEntriesForSet?: boolean;
+  },
+];
+export type MessageIds =
+  | 'objectKeysForMap'
+  | 'objectValuesForMap'
+  | 'objectEntriesForMap'
+  | 'objectKeysForSet'
+  | 'objectValuesForSet'
+  | 'objectEntriesForSet';
+
+export default createRule<Options, MessageIds>({
+  name: 'no-misused-object-likes',
+  defaultOptions: [
+    {
+      checkObjectKeysForMap: true,
+      checkObjectValuesForMap: true,
+      checkObjectEntriesForMap: true,
+      checkObjectKeysForSet: true,
+      checkObjectValuesForSet: true,
+      checkObjectEntriesForSet: true,
+    },
+  ],
+
+  meta: {
+    type: 'problem',
+    docs: {
+      description:
+        'Enforce check `Object.values(...)`, `Object.keys(...)`, `Object.entries(...)` usage with Map/Set objects',
+      requiresTypeChecking: false,
+    },
+    messages: {
+      objectKeysForMap: "Don't use `Object.keys()` for Map objects",
+      objectValuesForMap: "Don't use `Object.values()` for Map objects",
+      objectEntriesForMap: "Don't use `Object.entries()` for Map objects",
+      objectKeysForSet: "Don't use `Object.keys()` for Set",
+      objectValuesForSet: "Don't use `Object.values()` for Set",
+      objectEntriesForSet: "Don't use `Object.entries()` for Set",
+    },
+    schema: [
+      {
+        type: 'object',
+        additionalProperties: false,
+        properties: {
+          checkObjectKeysForMap: {
+            description: 'Check usage Object.keys for Map object',
+            type: 'boolean',
+          },
+          checkObjectValuesForMap: {
+            description: 'Check usage Object.values for Map object',
+            type: 'boolean',
+          },
+          checkObjectEntriesForMap: {
+            description: 'Check usage Object.entries for Map object',
+            type: 'boolean',
+          },
+          checkObjectKeysForSet: {
+            description: 'Check usage Object.keys for Set object',
+            type: 'boolean',
+          },
+          checkObjectValuesForSet: {
+            description: 'Check usage Object.values for Set object',
+            type: 'boolean',
+          },
+          checkObjectEntriesForSet: {
+            description: 'Check usage Object.entries for Set object',
+            type: 'boolean',
+          },
+        },
+      },
+    ],
+  },
+
+  create(context, [options]) {
+    const services = getParserServices(context);
+
+    function checkObjectMethodCall(
+      callExpression: TSESTree.CallExpression,
+    ): void {
+      const argument = callExpression.arguments[0];
+      const type = services.getTypeAtLocation(argument);
+      const argumentTypeName = type.getSymbol()?.name;
+      const callee = callExpression.callee as MemberExpression;
+      const objectMethod = (callee.property as Identifier).name;
+
+      if (argumentTypeName === 'Map') {
```

Comment:
[Bug] If a user defines their own `Map` class, this will falsely report:

```ts
class Map {
  value = 1;
}

const map = new Map();
Object.values(map);
```